pkg/client/restclient/request.go

```diff
@@ -786,22 +804,31 @@ func (r *Request) request(fn func(*http.Request, *http.Response)) error {
 	// TODO: Change to a timeout based approach.
 	maxRetries := 10
 	retries := 0
+	var req *http.Request
+	var resp *http.Response
 	for {
-		url := r.URL().String()
-		req, err := http.NewRequest(r.verb, url, r.body)
-		if err != nil {
-			return err
-		}
-		req.Header = r.headers
+		// TODO dedup multiple usage of WrapWithRetry in Stream/Watch/request methods
+		err := r.urlProvider.WrapWithRetry(
```

Comment:
I think I mentioned this a while ago, but I really don't like this pattern. This inverts the control of retries and hands it to the `urlProvider`. It doesn't respect the `maxRetries` defined above. And it's also not-integrated with the server-side rate-limiting (line 849 below). It looks to me that a RATE_LIMIT_EXCEEDED response from one server will cause the client to sleep, and then move on to a server that wasn't rate limiting.

I guess that is to say `urlProvider` should really only have one job: provide URLs. I would much rather see an implementation where a `urlProvider` provides a snapshot slice of URLs that are currently considered valid, and let this client do retry strategies/load balancing how it sees fit.